File: packages_py/fetch_auth_config/src/fetch_auth_config/resolution/env_resolver.py
import os
from typing import List, Optional, Union
from pydantic import BaseModel

class EnvVarResolveResult(BaseModel):
    value: Optional[str]
    source: Optional[str]
    tried: List[str]

def resolve_env_var(env_var_name: str) -> Optional[str]:
    """Look up a single environment variable value."""
    return os.getenv(env_var_name)

def resolve_env_var_chain(
    primary: Optional[str] = None,
    overwrite: Optional[Union[str, List[str]]] = None
) -> EnvVarResolveResult:
    """
    Try multiple env vars in order (overwrite -> primary -> fallbacks).
    Returns values, source, and list of tried vars.
    """
    tried: List[str] = []
    
    # 1. Try overwrite first (highest priority)
    if overwrite:
        overwrites = [overwrite] if isinstance(overwrite, str) else overwrite
        for env_var in overwrites:
            tried.append(env_var)
            val = resolve_env_var(env_var)
            if val is not None:
                return EnvVarResolveResult(value=val, source=env_var, tried=tried)

    # 2. Try primary
    if primary:
        tried.append(primary)
        val = resolve_env_var(primary)
        if val is not None:
            return EnvVarResolveResult(value=val, source=primary, tried=tried)

    return EnvVarResolveResult(value=None, source=None, tried=tried)

File: packages_mjs/fetch-auth-config/src/resolution/env-resolver.ts
export interface EnvVarChainConfig {
    /** Primary env var name (from env_api_key, etc.) */
    primary?: string;
    /** Overwrite env var (from overwrite_from_env) */
    overwrite?: string | string[];
}

export interface EnvVarResolveResult {
    value: string | null;
    source: string | null;
    tried: string[];
}

export function resolveEnvVar(
    envVarName: string
): { value: string | null; source: string | null } {
    const value = process.env[envVarName];
    return {
        value: value ?? null,
        source: value !== undefined ? envVarName : null
    };
}

export function resolveEnvVarChain(
    config: EnvVarChainConfig
): EnvVarResolveResult {
    const tried: string[] = [];

    // 1. Try overwrite first (highest priority)
    if (config.overwrite) {
        const overwrites = Array.isArray(config.overwrite)
            ? config.overwrite
            : [config.overwrite];
        for (const envVar of overwrites) {
            tried.push(envVar);
            const result = resolveEnvVar(envVar);
            if (result.value !== null) return { ...result, tried };
        }
    }

    // 2. Try primary
    if (config.primary) {
        tried.push(config.primary);
        const result = resolveEnvVar(config.primary);
        if (result.value !== null) return { ...result, tried };
    }

    return { value: null, source: null, tried };
}
